### PR TITLE
TensorRT dot product attention ops

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

DETR would failed if sequence length in MHA is large.
Add this op to implement core of mha by CUBLAS and CUDNN.
Note that we better merge this after #924 

## Modification

- Add custom ops `TRTDotProductAttention`.
- Add unit test.
- Add docs.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by AllentDan — COMMENTED** [review]
Visualization failed in my test. There were no objects detected. Here is the script I used:
```
python tools/deploy.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../mmdetection/configs/detr/detr_r50_8x2_150e_coco.py https://download.openmmlab.com/mmdetection/v2.0/detr/detr_r50_8x2_150e_coco/detr_r50_8x2_150e_coco_20201130_194835-2c4b8974.pth ../mmdetection/demo/demo.jpg --work-dir ../work_dirs/tensorrt/det/detr --device cuda --show
```
And here is the env of mine:
```
2022-08-30 15:16:15,958 - mmdeploy - INFO - 

2022-08-30 15:16:15,958 - mmdeploy - INFO - **********Environmental information**********
2022-08-30 15:16:16,087 - mmdeploy - INFO - sys.platform: linux
2022-08-30 15:16:16,087 - mmdeploy - INFO - Python: 3.8.12 (default, Oct 12 2021, 13:49:34) [GCC 7.5.0]
2022-08-30 15:16:16,087 - mmdeploy - INFO - CUDA available: True
2022-08-30 15:16:16,087 - mmdeploy - INFO - GPU 0: NVIDIA GeForce GTX 1660 SUPER
2022-08-30 15:16:16,087 - mmdeploy - INFO - CUDA_HOME: /usr/local/cuda
2022-08-30 15:16:16,087 - mmdeploy - INFO - NVCC: Cuda compilation tools, release 11.3, V11.3.58
2022-08-30 15:16:16,087 - mmdeploy - INFO - GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
2022-08-30 15:16:16,087 - mmdeploy - INFO - PyTorch: 1.10.2
2022-08-30 15:16:16,087 - mmdeploy - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.2, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

2022-08-30 15:16:16,087 - mmdeploy - INFO - TorchVision: 0.11.3
2022-08-30 15:16:16,087 - mmdeploy - INFO - OpenCV: 4.5.4
2022-08-30 15:16:16,087 - mmdeploy - INFO - MMCV: 1.5.0
2022-08-30 15:16:16,087 - mmdeploy - INFO - MMCV Compiler: GCC 7.5
2022-08-30 15:16:16,087 - mmdeploy - INFO - MMCV CUDA Compiler: 11.3
2022-08-30 15:16:16,087 - mmdeploy - INFO - MMDeploy: 0.7.0+a9c3d92
2022-08-30 15:16:16,087 - mmdeploy - INFO - 

2022-08-30 15:16:16,088 - mmdeploy - INFO - **********Backend information**********
2022-08-30 15:16:16,419 - mmdeploy - INFO - onnxruntime: 1.8.1  ops_is_avaliable : True
2022-08-30 15:16:16,446 - mmdeploy - INFO - tensorrt: 8.4.1.5   ops_is_avaliable : True
2022-08-30 15:16:16,459 - mmdeploy - INFO - ncnn: 1.0.20220830  ops_is_avaliable : True
2022-08-30 15:16:16,461 - mmdeploy - INFO - pplnn_is_avaliable: False
2022-08-30 15:16:16,486 - mmdeploy - INFO - openvino_is_avaliable: True
2022-08-30 15:16:16,505 - mmdeploy - INFO - snpe_is_available: False
2022-08-30 15:16:16,505 - mmdeploy - INFO - 

2022-08-30 15:16:16,505 - mmdeploy - INFO - **********Codebase information**********
2022-08-30 15:16:16,520 - mmdeploy - INFO - mmdet:      2.19.0
2022-08-30 15:16:16,520 - mmdeploy - INFO - mmseg:      0.25.0
2022-08-30 15:16:16,520 - mmdeploy - INFO - mmcls:      None
2022-08-30 15:16:16,520 - mmdeploy - INFO - mmocr:      None
2022-08-30 15:16:16,520 - mmdeploy - INFO - mmedit:     0.12.0
2022-08-30 15:16:16,520 - mmdeploy - INFO - mmdet3d:    None
2022-08-30 15:16:16,520 - mmdeploy - INFO - mmpose:     None
2022-08-30 15:16:16,521 - mmdeploy - INFO - mmrotate:   None
```

**Review by AllentDan — APPROVED** [review]
LGTM

File: tests/test_ops/test_ops.py [hunk 1]
```diff
@@ -1096,3 +1096,36 @@ def forward(self, x):
...
+
+    wrapped_model = WrapFunction(wrapped_function).eval()
+
+    with RewriterContext(
+            Config({'backend_config': {
+                'type': backend.backend_name
+            }}),
+            backend=backend.backend_name,
+            opset=11):
+        backend.run_and_validate(
```
Comment (tpoisonooo) [inline tests/test_ops/test_ops.py:1126]: This testcase seems not make sure the precision.

I think that we should construct a whole multiheadattention opr and compare the result with `torch.nn.MultiHeadAttention`.

Comment (grimoire) [reply tests/test_ops/test_ops.py:1121]: Done.
Precision will be tested inside [run_and_validate](https://github.com/open-mmlab/mmdeploy/blob/792c27b054b4a93f81d00e69a0d7f162b8fcafc0/tests/test_ops/utils.py#L158).

**Review by tpoisonooo — APPROVED** [review]
LGTM.

**Pull request merged**